ugvc/pipelines/mrd/pileup_based_read_features.py

```diff
@@ -0,0 +1,332 @@
+from __future__ import annotations
+
+import argparse
+import os
+import re
+import sys
+from collections import defaultdict
+from enum import Enum
+
+import pysam
+from simppl.simple_pipeline import SimplePipeline
+
+
+class VariantType(Enum):
+    SNV = 1
+    MONO_BASE_INDEL = 2
+    MULTI_BASE_INDEL = 3
+    UNIQUE_SNV = 4
+    UNIQUE_MONO_BASE_INDEL = 5
+    UNIQUE_MULTI_BASE_INDEL = 6
+    LOW_AF_SNV = 7
+    LOW_AF_MONO_BASE_INDEL = 8
+    LOW_AF_MULTI_BASE_INDEL = 9
+    COMMON_SNV = 10
+    COMMON_MONO_BASE_INDEL = 11
+    COMMON_MULTI_BASE_INDEL = 12
+
+
+def get_parser() -> argparse.ArgumentParser:
+    parser = argparse.ArgumentParser(prog="pileup_based_read_features", description="Extract pileup features per read")
+    parser.add_argument(
+        "--input_bam",
+        help="input bam file with aligned reads",
+        required=True,
+        type=str,
+    )
+    parser.add_argument(
+        "--input_bed",
+        help="bed_file specifying calling regions",
+        required=True,
+        type=str,
+    )
+    parser.add_argument(
+        "--ref",
+        help="reference genome",
+        required=True,
+        type=str,
+    )
+    parser.add_argument(
+        "--input_featuremap",
+        help="featuremap input to add new features to",
+        required=True,
+        type=str,
+    )
+    parser.add_argument(
+        "--output_pref",
+        help="output prefix",
+        required=True,
+        type=str,
+    )
+
+    return parser
+
+
+class PileupBasedReadFeatures:
+    FEATURES_LIST = [
+        VariantType.UNIQUE_SNV,
+        VariantType.UNIQUE_MONO_BASE_INDEL,
+        VariantType.UNIQUE_MULTI_BASE_INDEL,
+        VariantType.LOW_AF_SNV,
+        VariantType.LOW_AF_MONO_BASE_INDEL,
+        VariantType.LOW_AF_MULTI_BASE_INDEL,
+        VariantType.COMMON_SNV,
+        VariantType.COMMON_MONO_BASE_INDEL,
+        VariantType.COMMON_MULTI_BASE_INDEL,
+    ]
+
+    FEATURE_NAMES = [f.name for f in FEATURES_LIST]
+    MAX_REGION_SIZE = 100000
+    PADDING = 300
+    AF_THRESHOLD = 0.2
+
+    LAST_POSITION = 0
+    LAST_CHR = ""
+
+    def __init__(self, args):
```

Comment:
Wouldn't you want an interface you can invoke from another python function instead of only as a standalone script?

Reply:
You can invoke it from another python function:
 PileupBasedReadFeatures(args).process()
 
Do you mean you would want to specify the arguments explicitly?

Reply:
ok, done